### PR TITLE
Explore: Unification of logs/metrics/traces user interface

### PR DESCRIPTION
**Pull request by aocenas**
Design doc: https://docs.google.com/document/d/1hZURlhvU5LVejMWXeQTaTMOLAC73FQjQR_EjgYtOV9c/edit#heading=h.sp7xjuf6unar

### CODE REVIEW

File: packages/grafana-data/src/types/data.ts [hunk 1]
```diff
@@ -15,7 +15,7 @@ export enum LoadingState {
   Error = 'Error',
 }
 
-export type PreferredVisualisationType = 'graph' | 'table';
+export type PreferredVisualisationType = 'graph' | 'table' | 'logs' | 'trace';
```
Comment (aocenas) [inline packages/grafana-data/src/types/data.ts:18]: I think we can use this to figure out where to show the dataFrames. For data sources which won't use that we should just make sure the output is sensible (ie fallback to table table/graph)

File: public/app/features/explore/Explore.tsx [hunk 2]
```diff
@@ -332,7 +338,8 @@ export class Explore extends React.PureComponent<ExploreProps, ExploreState> {
               <SecondaryActions
                 addQueryRowButtonDisabled={isLive}
                 // We cannot show multiple traces at the same time right now so we do not show add query button.
-                addQueryRowButtonHidden={mode === ExploreMode.Tracing}
+                //TODO:unification
+                // addQueryRowButtonHidden={mode === ExploreMode.Tracing}
```
Comment (aocenas) [inline public/app/features/explore/Explore.tsx:342]: I think we can just show one trace view per frame and deal with multiple traces that way.

**Comment by aocenas** [comment]
We also need to check code like this one https://github.com/grafana/grafana/blob/72b8300571dc585041ffee6a35c82b79abeb686c/public/app/plugins/datasource/prometheus/datasource.ts#L221. It seems to be working but data sources should probably not be getting explore internal state and depend on that.

File: packages/grafana-data/src/types/data.ts (hunk 1 above)
Comment (torkelo) [reply packages/grafana-data/src/types/data.ts:18]: :+1:

File: public/app/features/explore/utils/ResultProcessor.ts [hunk 3]
```diff
@@ -16,23 +16,48 @@ import { getGraphSeriesModel } from 'app/plugins/panel/graph2/getGraphSeriesMode
...
+    for (const frame of this.dataFrames) {
+      if (isTimeSeries(frame, this.state.datasourceInstance?.meta.id)) {
+        if (shouldShowInVisualisationType(frame, 'graph')) {
+          this.graphFrames.push(frame);
+        }
+        if (shouldShowInVisualisationType(frame, 'table')) {
+          this.tableFrames.push(frame);
+        }
+      } else {
+        if (shouldShowInVisualisationType(frame, 'logs')) {
```
Comment (kaydelaney) [inline public/app/features/explore/utils/ResultProcessor.ts:43]: Do we need to nest these if/else-if statements inside the else statement starting at line 42 or can we just use all else-if statements instead?

Comment (aocenas) [reply public/app/features/explore/utils/ResultProcessor.ts:42]: the thing is time series by default are shown in both graph and table. So they need to be separate, other types are sort of priority based in the sense that when we think something should be logs we won't try to match this to anything else.

Comment (aocenas) [reply public/app/features/explore/utils/ResultProcessor.ts:42]: Also this assumes we will start using the meta attributes so may not work at the moment for all cases, but as I mentioned if we can do some sensible default that should be fine.

**Comment by kylebrandt** [comment]
For me this is a lot of JS :-) If I am to review (some concepts maybe?) I think I would need some context or a summary.

**Comment by aocenas** [comment]
@kylebrandt seems like you ware added because the codeowners config and this touches some backend files. I think the automatic code owners assignment isn't that great in some cases,

File: public/app/features/explore/QueryRow.tsx [hunk 4]
```diff
@@ -102,12 +100,13 @@ export class QueryRow extends PureComponent<QueryRowProps, QueryRowState> {
   };
 
   setReactQueryEditor = () => {
-    const { mode, datasourceInstance } = this.props;
+    const { datasourceInstance } = this.props;
     let QueryEditor;
 
-    if (mode === ExploreMode.Metrics && datasourceInstance.components?.ExploreMetricsQueryField) {
+    // TODO:unification
```
Comment (davkal) [inline public/app/features/explore/QueryRow.tsx:106]: Looks like there is no other way than removing this field specialisation hand add a switcher to the field itself.

File: public/app/features/explore/state/actions.ts [hunk 5]
```diff
@@ -451,23 +431,24 @@ export const runQueries = (exploreId: ExploreId): ThunkResult<void> => {
 
     // Some datasource's query builders allow per-query interval limits,
     // but we're using the datasource interval limit for now
-    const minInterval = datasourceInstance.interval;
+    const minInterval = datasourceInstance?.interval;
```
Comment (davkal) [inline public/app/features/explore/state/actions.ts:434]: Why would there ever not be an instance?

File: public/app/features/explore/utils/ResultProcessor.ts [hunk 6]
```diff
@@ -16,27 +15,51 @@ import { getGraphSeriesModel } from 'app/plugins/panel/graph2/getGraphSeriesMode
...
+      } else if (isTimeSeries(frame, this.state.datasourceInstance?.meta.id)) {
+        if (shouldShowInVisualisationType(frame, 'graph')) {
+          this.graphFrames.push(frame);
+        }
+        if (shouldShowInVisualisationType(frame, 'table')) {
+          this.tableFrames.push(frame);
+        }
+      } else {
+        // We fallback to table if we do not have any better meta info about the dataframe.
+        this.tableFrames.push(frame);
```
Comment (davkal) [inline public/app/features/explore/utils/ResultProcessor.ts:51]: 👍

File: public/app/plugins/datasource/elasticsearch/datasource.ts [hunk 7]
```diff
@@ -363,7 +363,7 @@ export class ElasticDatasource extends DataSourceApi<ElasticsearchQuery, Elastic
       let queryObj;
       if (target.isLogsQuery || queryDef.hasMetricOfType(target, 'logs')) {
         target.bucketAggs = [queryDef.defaultBucketAgg()];
-        target.metrics = [queryDef.defaultMetricAgg()];
+        target.metrics = [];
```
Comment (davkal) [inline public/app/plugins/datasource/elasticsearch/datasource.ts:366]: Ah ok, we're not requesting the aggregation anymore.

File: public/app/plugins/datasource/elasticsearch/module.ts [hunk 8]
```diff
@@ -11,5 +10,4 @@ class ElasticAnnotationsQueryCtrl {
 export const plugin = new DataSourcePlugin(ElasticDatasource)
   .setQueryCtrl(ElasticQueryCtrl)
   .setConfigEditor(ConfigEditor)
-  .setExploreLogsQueryField(ElasticsearchQueryField)
```
Comment (davkal) [inline public/app/plugins/datasource/elasticsearch/module.ts:14]: The component file could be removed?

File: public/app/plugins/datasource/influxdb/module.ts [hunk 9]
```diff
@@ -16,5 +15,4 @@ export const plugin = new DataSourcePlugin(InfluxDatasource)
   .setConfigEditor(ConfigEditor)
   .setQueryCtrl(InfluxQueryCtrl)
   .setAnnotationQueryCtrl(InfluxAnnotationsQueryCtrl)
-  .setExploreLogsQueryField(InfluxLogsQueryField)
```
Comment (davkal) [inline public/app/plugins/datasource/influxdb/module.ts:19]: Remove component?

File: public/app/plugins/datasource/jaeger/datasource.ts [hunk 10]
```diff
@@ -49,6 +49,9 @@ export class JaegerDatasource extends DataSourceApi<JaegerQuery> {
                     values: response?.data?.data || [],
                   },
                 ],
+                meta: {
+                  preferredVisualisationType: 'trace',
```
Comment (davkal) [inline public/app/plugins/datasource/jaeger/datasource.ts:53]: I feel like these values should be constants from something in core grafana.

File: public/app/plugins/datasource/loki/datasource.ts [hunk 11]
```diff
@@ -93,30 +91,7 @@ export class LokiDatasource extends DataSourceApi<LokiQuery, LokiOptions> {
...
-                );
-              } else {
-                return dataQueryResponse;
-              }
-            })
-          )
-        )
-      );
-    }
+    filteredTargets.forEach(target => subQueries.push(this.runRangeQuery(target, options, filteredTargets.length)));
```
Comment (davkal) [inline public/app/plugins/datasource/loki/datasource.ts:94]: We're dropping the instant query here. Any way we can have it back for metric queries (or we just sent it anyway for logs and see what happens/deal with it in the response).

File: public/app/plugins/datasource/loki/datasource.ts [hunk 12]
```diff
@@ -150,7 +125,10 @@ export class LokiDatasource extends DataSourceApi<LokiQuery, LokiOptions> {
       filter((response: any) => (response.cancelled ? false : true)),
       map((response: { data: LokiResponse }) => {
         if (response.data.data.resultType === LokiResultType.Stream) {
-          throw new Error('Metrics mode does not support logs. Use an aggregation or switch to Logs mode.');
+          return {
+            data: [],
+            key: `${target.refId}_instant`,
```
Comment (davkal) [inline public/app/plugins/datasource/loki/datasource.ts:130]: Unclear what's happening here. Also, the `_instant` suffix suggests this has something to do with instant queries.

**Review by davkal — APPROVED** [review]
Awesome 🚀 

No blockers from my side.

Minor:

- Loki is missing the table result. Either add in this PR if simple, or add a comment TODO
- Elastic agg histogram from the datasource was nice. Is there no other way to reintroduce this?

**Comment by kaydelaney** [comment]
Failing right now as it depends on https://github.com/grafana/grafana-plugin-sdk-go/pull/191 which needs to get merged first

File: packages/grafana-data/src/types/data.ts [hunk 13]
```diff
@@ -47,6 +47,7 @@ export interface QueryResultMeta {
   searchWords?: string[]; // used by log models and loki
   limit?: number; // used by log models and loki
   json?: boolean; // used to keep track of old json doc values
+  instant?: boolean;
```
[the remark this answers is not in the document]
Comment (aocenas) [reply packages/grafana-data/src/types/data.ts:50]: Do we still need this? I thought that was solved by the preferredVis type?

File: public/app/features/explore/Explore.tsx [hunk 14]
```diff
@@ -334,7 +340,8 @@ export class Explore extends React.PureComponent<ExploreProps, ExploreState> {
               <SecondaryActions
                 addQueryRowButtonDisabled={isLive}
                 // We cannot show multiple traces at the same time right now so we do not show add query button.
-                addQueryRowButtonHidden={mode === ExploreMode.Tracing}
+                //TODO:unification
```
[the remark this answers is not in the document]
Comment (aocenas) [reply public/app/features/explore/Explore.tsx:343]: My idea for simple solution would be to just show 1 trace view per each query. Would not be very practical from user perspective but would be easy to implement without any special casing here.

File: packages/grafana-data/src/types/data.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (kaydelaney) [reply packages/grafana-data/src/types/data.ts:50]: Yeah seems like it's not used anymore

File: public/app/features/explore/state/actions.ts (hunk 5 above)
Comment (kaydelaney) [reply public/app/features/explore/state/actions.ts:434]: Good question, but it's defined as an optional property so the `?` is needed to suppress a strict null error.

**Comment by NotAFile** [comment]
Is it intended that this design doc is private?

**Pull request merged**